### PR TITLE
Classification Fixes

### LINKED ISSUE #104: Wrong Output Size for Classification Problems
**Describe the bug**
When doing multi-class classification in the current master branch, and the CSV file have the class labels in single column, the number of classes are seen as 1.

**To Reproduce**
1. Checkout to master branch
2. Use the config file attached and prepare the csv file based on the images I attached.
3. Run gandlf

**Expected behavior**
image

I think the problem is due to the populate_header_in_parameters function. There is an if statement there that setting the number of classes to the number of prediction headers, which is one header, consisting of the image labels.
image

**Screenshots**
[config_class.txt](https://github.com/CBICA/GaNDLF/files/6710191/config_class.txt)

ages.githubusercontent.com/24361544/122981569-31de2b00-d39
[mini_training_data - Copy.txt](https://github.com/CBICA/GaNDLF/files/6710203/mini_training_data.-.Copy.txt)
a-11eb-9ba0-4872c0b1079f.png)
**GaNDLF Version**
0.0.10

**GaNDLF Version**
0.0.10

**Desktop (please complete the following information):**
 - Ubuntu 20.04 WSL 2

### PR DESCRIPTION
**Pull request by orhunguley**
Fixes #104 

## Proposed Changes

- Fixes the number of classes issue
- Categorical Cross Entropy is added as loss and metric.
- Fixes an issue occuring during validation phase of classification problems due to wrong label shape. 

## Checklist

- [x] I have read the [`CONTRIBUTING`](https://github.com/CBICA/GaNDLF/blob/master/CONTRIBUTING.md) guide
- [x] My PR is based from the [current GaNDLF master ](https://garygregory.wordpress.com/2016/11/10/how-to-catch-up-my-git-fork-to-master/)
- [ ] Non-breaking change (would not break existing functionality): if changes breaks current code, please provide as many details as possible --> not exactly sure if the fix at populate_header_in_parameters() would break anything or not.

- [ ] Function/class source code documentation added/updated
- [ ] Code has been [blacked](https://github.com/psf/black#usage) for style consistency
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/CBICA/GaNDLF/blob/master/GANDLF/version.py)
- [ ] If adding a submodule, add to list of expections for black styling in [pyproject.toml](https://github.com/CBICA/GaNDLF/blob/master/pyproject.toml) file 
- [ ] [Usage documentation](https://github.com/CBICA/GaNDLF/blob/master/docs) has been updated, if appropriate
- [ ] [History](https://github.com/CBICA/GaNDLF/blob/master/HISTORY.md) has been updated, if appropriate
- [ ] Tests added or modified to cover the changes
- [ ] Integration tests passed locally by running `pytest`

### CODE REVIEW

**Comment by sarthakpati** [comment]
Hey @orhunguley,

This is a great PR! I wanted to give you a heads up that the `test_inference_classification_rad_3d` is failing, though... 

Cheers!

File: GANDLF/utils.py [hunk 1]
```diff
@@ -316,10 +316,13 @@ def populate_header_in_parameters(parameters, headers):
     # initialize common parameters based on headers
     parameters["headers"] = headers
     # ensure the number of output classes for model prediction is working correctly
-    if len(headers["predictionHeaders"]) > 0:
+
+    if len(headers["predictionHeaders"]) > 1:
```
Comment (sarthakpati) [inline GANDLF/utils.py:320]: Can you give me the rationale behind why the comparison value is `1` and not `0`?

File: GANDLF/utils.py [hunk 2]
```diff
@@ -316,10 +316,13 @@ def populate_header_in_parameters(parameters, headers):
...
         parameters["model"]["num_classes"] = len(headers["predictionHeaders"])
     else:
         parameters["model"]["num_classes"] = len(parameters["model"]["class_list"])
+    
+    parameters["model"]["num_classes"] = len(parameters["model"]["class_list"])
```
Comment (sarthakpati) [inline GANDLF/utils.py:325]: So, I would probably do something like this:
```suggestion
    is_regression, _, _ = find_problem_type(parameters["headers"], parameters["model"]["final_layer"])
    
    # if the problem type is classification/segmentation, ensure the number of classes are picked from the configuration
    if not is_regression:
        parameters["model"]["num_classes"] = len(parameters["model"]["class_list"])
```

Thoughts?

**Review by sarthakpati — CHANGES_REQUESTED** [review]
Hey @orhunguley,

This is a great PR! I wanted to give you a heads up that the test_inference_classification_rad_3d is failing, though...

Also, I would also run `black` on the project. 😄 

Cheers!

File: GANDLF/training_loop.py [hunk 3]
```diff
@@ -85,12 +85,13 @@ def step(model, image, label, params):
...
         output = model(image)
-    
-    
+
     if "medcam_enabled" in params and params["medcam_enabled"]:
         output, attention_map = output
 
-    # print("Output shape : ", output.shape, flush=True)
+    print("Output shape : ", output.shape, flush=True)
+    print("label shape : ", label.shape, flush=True)
```
Comment (sarthakpati) [inline GANDLF/training_loop.py:93]: If this is something for debugging, please remove, or if this needs to be presented to the user, please under verbose flag like this:

```suggestion
    if params["verbose"]:
        print("Output shape : ", output.shape, flush=True)
        print("label shape : ", label.shape, flush=True)
```

Comment (orhunguley) [reply GANDLF/training_loop.py:93]: Thanks @sarthakpati! it was for debugging and I will remove completely

**Comment by sarthakpati** [comment]
Keeping this PR as draft until the debugging is done. 😄

**Comment by sarthakpati** [comment]
Once the [codacy suggestions](https://app.codacy.com/gh/CBICA/GaNDLF/pullRequest?prid=7692832) are in, we should be good to merge!

**Review by sarthakpati — APPROVED** [review]
LGTM!

**Pull request merged**